### PR TITLE
i#6831 sched refactor, step 5: Split check for whether to switch inputs

### PR DESCRIPTION
**Pull request by derekbruening**
Splits the mode-specific middle of next_record() into a new virtual method check_for_input_switch() for identifying whether to trigger a switch and call pick_next_input_for_mode().  The new method is implemented in the new subclasses scheduler_{dynamic,fixed,replay}.cpp.

Issue: #6831

### CODE REVIEW

File: clients/drcachesim/scheduler/scheduler_impl.h [hunk 1]
```diff
@@ -1032,11 +1049,18 @@ class scheduler_replay_tmpl_t : public scheduler_impl_tmpl_t<RecordType, ReaderT
...
         typename scheduler_impl_tmpl_t<RecordType, ReaderType>::schedule_record_t;
+    using input_info_t =
+        typename scheduler_impl_tmpl_t<RecordType, ReaderType>::input_info_t;
 
 protected:
     stream_status_t
     pick_next_input_for_mode(output_ordinal_t output, uint64_t blocked_time,
                              input_ordinal_t prev_index, input_ordinal_t &index) override;
+    stream_status_t
+    process_next_record_candidate(output_ordinal_t output, RecordType &record,
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h:1060]: Renamed to "check_for_input_switch".  That is the purpose: identify whether we should switch inputs.

File: clients/drcachesim/scheduler/scheduler_impl.h [hunk 2]
```diff
@@ -763,6 +763,18 @@ template <typename RecordType, typename ReaderType> class scheduler_impl_tmpl_t
...
     }
 
+    // Helper for pick_next_input() specialized by mapping_t mode.
+    virtual stream_status_t
+    process_next_record_candidate(output_ordinal_t output, RecordType &record,
+                                  input_info_t *input, uint64_t cur_time,
+                                  bool &need_new_input, bool &preempt,
+                                  uint64_t &blocked_time)
+    {
+        // Return an error, rather than being pure virtual, to make subclassing
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h:773]: Made it pure-virtual and overrode in the test classes: there are just 2 of them.

File: clients/drcachesim/scheduler/scheduler_impl.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h]: Added documentation.

File: clients/drcachesim/scheduler/scheduler_impl.h [hunk 3]
```diff
@@ -758,7 +758,10 @@ template <typename RecordType, typename ReaderType> class scheduler_impl_tmpl_t
     pick_next_input_for_mode(output_ordinal_t output, uint64_t blocked_time,
                              input_ordinal_t prev_index, input_ordinal_t &index) = 0;
 
-    // Helper for pick_next_input() specialized by mapping_t mode.
+    // Helper for pick_next_input() specialized by mapping_t mode: determines
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler_impl.h:761]: Also add when the derived class will be allowed to determine whether to check for input switch.

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/scheduler/scheduler_impl.h (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h:761]: Oh sorry that new comment is on pick_next_input_for_mode().  What we want here is that this is called on every record, inside next_record().  Added that (and s/pick_next_input/next_record/).

Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h:761]: Further added to link the other way:
  
If this returns true for "need_new_input", pick_next_input_for_mode() is called.

Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler_impl.h]: Updated the PR title and description as well.

**Pull request merged**